### ISSUE TITLE
No callback on Branch.io failure

### ISSUE THREAD

**Issue by ovidiu-anghelina** [blackmenthor/flutter-branch-io#13]
In the case of the Branch.io library failing to get a response from the Branch.io back-end, your plugin does not trigger anything and as such the app does not know that something went wrong with Branch.

For example, in the case of lack of internet connection, the following gets printed in the logs when calling `setupBranchIO`: `I/FlutterBranchIo(24473): Trouble initializing Branch.  Branch API Error: poor network connectivity. Please try again later.`. The listening stream is not notified in any way however, so the app is completely unaware of this and cannot react appropriately.

My temporary workaround to this issue is a hardcoded 5 second timer - if 5 seconds have passed without getting any value from the listening stream, I assume Branch failed to connect and display a popup that allows retrying.

I don't know if there are other potential failures that Branch can encounter, but they should all be handled and fed back to the listening stream. It could be intermittent internet connection, or Branch.io back-end downtime (happens rarely), or just a very slow internet connection that times out - all of which should be handled by your plugin.

My suggestion is to change `listenToDeepLinkStream` to return a `Stream<Map<String, String>>` instead of `Stream<String>`. This would allow you to return different key-value pairs, with keys such as "errorCode" for different errors that could occur, "data" for the successful JSON string, or "status" to inform that the listener has been attached ([for this other issue here](https://github.com/blackmenthor/flutter-branch-io/issues/11#issuecomment-496954060))

**Comment by blackmenthor** [comment]
alright i'll work on this as soon as i have time, i saw your other comments too on the race condition and will take a look on that as well. So sorry for the delay, been busy recently

**Comment by ovidiu-anghelina** [comment]
Given that I needed this sooner, I created my own implementation and found out that a Stream of Map<String, String> would be overengineered because the Flutter channels support error handling - it's just that it was never used in your implementation.

As for the race condition, I have made it so that on Android, when the event channel listener is attached on the native side, it also triggers branch.io - this way when the Flutter app initializes you only need to attach the listener and it will get the data as well. You still need the other channel for 'manually' triggering Branch.io when the activity lifecycle changes (or if you want the user to tap a 'Retry' button after an error in the event stream).
